pkg/schedulingtypes/jobscheduler.go

```diff
@@ -0,0 +1,229 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package schedulingtypes
+
+import (
+	"fmt"
+
+	"github.com/kubernetes-sigs/federation-v2/pkg/apis/core/typeconfig"
+	fedschedulingv1a1 "github.com/kubernetes-sigs/federation-v2/pkg/apis/scheduling/v1alpha1"
+	fedclientset "github.com/kubernetes-sigs/federation-v2/pkg/client/clientset/versioned"
+	"github.com/kubernetes-sigs/federation-v2/pkg/controller/util"
+	. "github.com/kubernetes-sigs/federation-v2/pkg/controller/util"
+	"github.com/kubernetes-sigs/federation-v2/pkg/controller/util/planner"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	pkgruntime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+)
+
+const (
+	JSPKind          = "JobSchedulingPreference"
+	ParallelismField = "parallelism"
+	CompletionsField = "completions"
+	ParallelismPath  = "spec.parallelism"
+	CompletionsPath  = "spec.completions"
+)
+
+func init() {
+	schedulingType := SchedulingType{
+		Kind:             JSPKind,
+		SchedulerFactory: NewJobScheduler,
+	}
+	RegisterSchedulingType("jobs.batch", schedulingType)
+}
+
+type JobScheduler struct {
+	plugin *Plugin
+
+	fedClient fedclientset.Interface
+
+	controllerConfig *ControllerConfig
+
+	eventHandlers SchedulerEventHandlers
+}
+
+func NewJobScheduler(controllerConfig *ControllerConfig, eventHandlers SchedulerEventHandlers) (Scheduler, error) {
+	fedClient, _, _ := controllerConfig.AllClients("job-scheduler")
+	scheduler := &JobScheduler{
+		controllerConfig: controllerConfig,
+		eventHandlers:    eventHandlers,
+		fedClient:        fedClient,
+	}
+
+	return scheduler, nil
+}
+
+func (j *JobScheduler) Kind() string {
+	return RSPKind
+}
+
+func (j *JobScheduler) StartPlugin(typeConfig typeconfig.Interface, stopChan <-chan struct{}) error {
+	kind := typeConfig.GetTemplate().Kind
+	plugin, err := NewPlugin(j.controllerConfig, j.eventHandlers, typeConfig)
+	if err != nil {
+		return fmt.Errorf("Failed to initialize replica scheduling plugin for %q: %v", kind, err)
+	}
+	plugin.Start(stopChan)
+	j.plugin = plugin
+
+	return nil
+}
+
+func (j *JobScheduler) ObjectType() pkgruntime.Object {
+	return &fedschedulingv1a1.JobSchedulingPreference{}
+}
+
+func (j *JobScheduler) FedList(namespace string, options metav1.ListOptions) (pkgruntime.Object, error) {
+	return j.fedClient.SchedulingV1alpha1().JobSchedulingPreferences(j.controllerConfig.TargetNamespace).List(options)
+}
+
+func (j *JobScheduler) FedWatch(namespace string, options metav1.ListOptions) (watch.Interface, error) {
+	return j.fedClient.SchedulingV1alpha1().JobSchedulingPreferences(j.controllerConfig.TargetNamespace).Watch(options)
+}
+
+func (j *JobScheduler) Start(stopChan <-chan struct{}) {
```

Comment:
(No action required) I find it pretty confusing that `Start` is a method of the interface that isn't used.  It suggests a future refactor.